gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/Orchestrator.java

```diff
@@ -229,33 +229,39 @@ public void orchestrate(Spec spec, Properties jobProps, long triggerTimestampMil
-                flowName, flowMetadata);
-
-        if (!compiledDagOptional.isPresent()) {
-          Instrumented.markMeter(this.flowOrchestrationFailedMeter);
-          return;
-        }
-        Dag<JobExecutionPlan> compiledDag = compiledDagOptional.get();
-        if (compiledDag.isEmpty()) {
-          FlowCompilationValidationHelper.populateFlowCompilationFailedEventMessage(eventSubmitter, flowSpec, flowMetadata);
-          Instrumented.markMeter(this.flowOrchestrationFailedMeter);
+        try {
+          TimingEvent flowCompilationTimer = new TimingEvent(this.eventSubmitter, TimingEvent.FlowTimings.FLOW_COMPILED);
+          Map<String, String> flowMetadata = TimingEventUtils.getFlowMetadata(flowSpec);
+          Optional<Dag<JobExecutionPlan>> compiledDagOptional =
+              this.flowCompilationValidationHelper.validateAndHandleConcurrentExecution(flowConfig, flowSpec, flowGroup,
+                  flowName, flowMetadata);
+
+          if (!compiledDagOptional.isPresent()) {
+            Instrumented.markMeter(this.flowOrchestrationFailedMeter);
+            return;
+          }
+          Dag<JobExecutionPlan> compiledDag = compiledDagOptional.get();
+          if (compiledDag.isEmpty()) {
+            FlowCompilationValidationHelper.populateFlowCompilationFailedEventMessage(eventSubmitter, flowSpec,
+                flowMetadata);
+            Instrumented.markMeter(this.flowOrchestrationFailedMeter);
+            sharedFlowMetricsSingleton.conditionallyUpdateFlowGaugeSpecState(spec,
+                SharedFlowMetricsSingleton.CompiledState.FAILED);
+            _log.warn("Cannot determine an executor to run on for Spec: " + spec);
+            return;
+          }
           sharedFlowMetricsSingleton.conditionallyUpdateFlowGaugeSpecState(spec,
-              SharedFlowMetricsSingleton.CompiledState.FAILED);
-          _log.warn("Cannot determine an executor to run on for Spec: " + spec);
-          return;
-        }
-        sharedFlowMetricsSingleton.conditionallyUpdateFlowGaugeSpecState(spec,
-            SharedFlowMetricsSingleton.CompiledState.SUCCESSFUL);
+              SharedFlowMetricsSingleton.CompiledState.SUCCESSFUL);
 
-        FlowCompilationValidationHelper.addFlowExecutionIdIfAbsent(flowMetadata, compiledDag);
-        flowCompilationTimer.stop(flowMetadata);
+          FlowCompilationValidationHelper.addFlowExecutionIdIfAbsent(flowMetadata, compiledDag);
+          flowCompilationTimer.stop(flowMetadata);
 
-        // Depending on if DagManager is present, handle execution
-        submitFlowToDagManager(flowSpec, compiledDag);
+          // Depending on if DagManager is present, handle execution
+          submitFlowToDagManager(flowSpec, compiledDag);
+        } finally {
+          // remove from the flow catalog, regardless of whether the flow was successfully validated and permitted to exec (concurrently)
+          this.dagManager.removeFlowSpecIfAdhoc(flowSpec);
+        }
```

Comment:
too bad the diff above doesn't clearly indicate it was solely an indentation change to add the `try ... finally` here.  the purpose of which is to ensure FlowCatalog cleanup, come what may